ch4/rma: enable fallback for win allocate when shm allocation fails

minsii opened the pull request:
<h3>Background</h3>

In CH4 win_allocate, we use the following logic for memory allocation:
1. Check if we can use global symmetric heap with shm allocation (i.e., page aligned size, or noncontig enabled)
2. If *1* is TRUE, try global symmetric heap with shm memory on multi-process node (or mmap on single-process node)
3. If *1* is FALSE or *2* returns a mapping failure (currently we only handles the case that no available starting address after 100 times retry, other failures are handled as mpi error):
   3.a. Use shm memory allocation on the node with multiple processes
   3.b. Use malloc on the node with single process

<h3>Problem</h3>

- For `MPI_Win_allocate`, shm allocation is only a performance optimization in MPICH but not required by standard. Thus, we should provide a fallback routine when both *2* and *3.a* fails, i.e., use malloc.

- For `MPI_Win_allocate_shared`, it requires shared memory as defined in standard, thus we still returns MPI error if shm allocation fails.

<h3>Changes in this PR</h3>

- Change `MPIDI_CH4U_allocate_shm_segment` in order to handle all MPL* shm allocation failing internally.
  + If MPL error is reported by any of the local process, this routine internally cleans up allocated resource and set output parameter `mapfail_flag=true` with `MPI_SUCCESS` as return value;
  + If MPI communication error occurs, still return MPI error.

- Change `MPIDI_CH4I_allocate_symshm_segment` for global symmetric heap allocation. Previously we only handle map failing (i.e., invalid start address passed to `mmap` with `MAP_FIXED` flag) as non-MPI error. This PR changes it as below:
  + Similar to `MPIDI_CH4U_allocate_shm_segment`, handles all MPL shm error as non-MPI error (return `MPI_SUCCESS`) and cleans up allocated resource if such error occurs. The output parameter is `map_result`, with value `SUCCESS|MAP_FAIL|OTHER_FAIL`. `MAP_FAIL` indicates the error caused by *invalid start address*, and `OTHER_FAIL` covers any other failure (e.g., *opened too many files*)
  + In the caller routine `MPIDI_CH4R_get_shm_symheap`, if `MAP_FAIL` returns then retry with another start address; if `OTHER_FAIL` returns, which means that node can no longer allocate shm segments, then it stops retrying and return with `fail_flag=true`.

- Changes in `MPIDI_CH4I_win_shm_alloc_impl`:
  + Allows the caller to pass `bool shm_required` parameter (`false` by `win_allocate`, and `true` by `win_allocate_shared`)
  + When both global symheap and shm allocation fail, if `shm_required==false` then use `malloc`; otherwise generates an MPI error.

- Changes in `MPIDI_POSIX_mpi_win_allocate_hook`:
  + If shm segment is not allocated (i.e., `mmap_addr == NULL`), do not initialize shm mutex and do not set `MPIDI_CH4U_WIN(win, shm_allocated)` (this disables load/store RMA operation)

- Changes in both `MPIDI_POSIX_mpi_win_allocate_hook` and `MPIDI_POSIX_mpi_win_allocate_shared_hook`:
  + If shm mutex allocation fails, do not set `MPIDI_CH4U_WIN(win, shm_allocated)=1` (this disables load/store RMA operation).

minsii commented:
test:jenkins/ch4/most

minsii commented:
test:jenkins/ch4/most

minsii commented:
test:jenkins/ch4/most

hajimefu commented:
test:intel:ch4/ofi/sockets/gnu/force-am-disabled,force-am-enabled/auto,netmod,no-odd-even,thread/debug,default

hajimefu commented on src/mpid/ch4/shm/posix/posix_win.h:
```diff
@@ -409,27 +409,31 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_POSIX_mpi_win_allocate_hook(MPIR_Win * win)
-     * more than 1 processes exist on the node. */
-    if (shm_comm_ptr == NULL || !MPL_proc_mutex_enabled())
+    /* Enable shm RMA only when interprocess mutex is supported,
+     * more than 1 processes exist on the node, and shm memory has been successfully allocated. */
```
"shm memory" sounds weird (redundant) to me. "shm", "shmem", or "shared memory"?

minsii replied:
Changed to `shm buffer`, as it indicates the window buffer.

hajimefu commented on src/mpid/ch4/shm/posix/posix_win.h:
```diff
@@ -409,27 +409,31 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_POSIX_mpi_win_allocate_hook(MPIR_Win * win)
-    if (shm_comm_ptr == NULL || !MPL_proc_mutex_enabled())
+    /* Enable shm RMA only when interprocess mutex is supported,
+     * more than 1 processes exist on the node, and shm memory has been successfully allocated. */
+    if (shm_comm_ptr == NULL || !MPL_proc_mutex_enabled() || !MPIDI_CH4U_WIN(win, mmap_addr))
```
This is a bit off topic but I think this line in MPL configure
https://github.com/pmodels/mpich/blob/master/src/mpl/configure.ac#L795
should be
```
if test "$ac_cv_func_pthread_mutexattr_setpshared" = "yes" ; then
```
instead of "no". This would affect the behavior of `MPL_proc_mutex_enabled()`, so if we are to fix it, I think it's worth including it in this PR and test it together.

hajimefu replied:
Also, should this change be really included in this patch?

minsii replied:
I think it makes sense to have the change in `posix_win` in this commit, as it is a part of the fallback routine when shm fails at win allocate (i.e., it disables load/store based RMA when shm allocation fails).

Let me fix the `ac_cv_func_pthread_mutexattr_setpshared` issue in a separate commit within this PR.

minsii replied:
After check again, I actually found that the following manual macro definition is unneeded. `AC_CHECK_FUNCS` will define it automatically.
```
    if test "$ac_cv_func_pthread_mutexattr_setpshared" = "no" ; then
      AC_DEFINE(HAVE_PTHREAD_MUTEXATTR_SETPSHARED,1, [Define if pthread_mutexattr_setpshared is available.])
    fi
```

hajimefu commented on src/mpid/ch4/src/ch4_impl.h:
```diff
@@ -954,47 +954,73 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_CH4U_wait_am_acc(MPIR_Win * win, int target_r
-        /* wait for other processes to attach to win */
-        mpi_errno = MPIR_Barrier(shm_comm_ptr, &errflag);
-        MPIR_ERR_CHKANDJUMP(errflag, mpi_errno, MPI_ERR_OTHER, "**coll_fail");
+        if (!shm_fail_flag) {
```
```C
if (shm_fail_flag)
    goto map_fail;

/* handle success case here */
```
would improve code readability.

minsii replied:
done.

hajimefu commented on src/mpid/ch4/src/ch4r_symheap.h:
```diff
@@ -222,71 +235,99 @@ static inline int MPIDI_CH4I_allocate_symshm_segment(MPIR_Comm * shm_comm_ptr,
-            MPIR_ERR_CHKANDJUMP(mpl_err != MPL_SHM_SUCCESS, mpi_errno, MPI_ERR_OTHER,
-                                "**alloc_shar_mem");
-
+        if (map_result == MPIDI_CH4I_SYMSHM_SUCCESS) {
```
Again, I would prefer
```C
if (map_result != MPIDI_CH4I_SYMSHM_SUCCESS)
     goto map_fail;
...
```
for readability.

minsii replied:
fixed

hajimefu commented on src/mpid/ch4/src/ch4r_symheap.h:
```diff
@@ -222,71 +235,99 @@ static inline int MPIDI_CH4I_allocate_symshm_segment(MPIR_Comm * shm_comm_ptr,
-            MPIR_ERR_CHKANDJUMP((mpl_err != MPL_SHM_EINVAL && mpl_err != MPL_SHM_SUCCESS),
-                                mpi_errno, MPI_ERR_OTHER, "**alloc_shar_mem");
+      result_sync:
+        /* check results of all processes, MAP_FAIL if no OTHER_FAIL is reported. */
```
"MAP_FAIL if no OTHER_FAIL is reported" -- I don't understand what this means. The code looks like "MAP_FAIL if any process reports failure".

minsii replied:
I improved the comment. It should make it clear.
```
        /* check results of all processes. If any failure happens (max result > 0),
         * return MPIDI_CH4I_SYMSHM_OTHER_FAIL if anyone reports it (max result == 2).
         * Otherwise return MPIDI_CH4I_SYMSHM_MAP_FAIL (max result == 1). */
```

hajimefu commented on src/mpid/ch4/src/ch4r_symheap.h:
```diff
@@ -450,38 +493,31 @@ static inline int MPIDI_CH4R_get_shm_symheap(MPI_Aint shm_size, MPI_Aint * shm_o
+                mpi_errno = MPIDI_CH4U_destroy_shm_segment(mapsize, shm_segment_hdl_ptr, base_ptr);
+                if (mpi_errno)
+                    MPIR_ERR_POP(mpi_errno);
+            } else if (map_result == MPIDI_CH4I_SYMSHM_SUCCESS) {
```
Can this block be:
```C
if (all_map_result != MPIDI_CH4I_SYMSHM_SUCCESS && mapsize > 0 && map_result == MPIDI_CH4I_SYMSHM_SUCCESS) {
     if (shm_comm_ptr != NULL) {
     } else {
     }
}
```
(no logic change, just a style preference.)

minsii replied:
It makes sense. fixed.

hajimefu commented on src/mpid/ch4/netmod/ofi/ofi_win.h:
```diff
@@ -1240,6 +1240,10 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_NM_mpi_win_free_hook(MPIR_Win * win)
 
     if (MPIDI_OFI_ENABLE_RMA) {
+        /* ofi window is not initialized */
+        if (!MPIDI_OFI_WIN(win).win_id)
```
Is it safe to assume that valid `win_id` is always non-zero? (even after applying #3493?)

minsii replied:
It looks like the new `win_id` can be zero. Do you have any suggestion to figure out whether ofi_win is initialized ? I could not find any existing parameter that we can reuse.

An alternate option would be to add an `init_status` in `MPIDI_Devwin_t` to indicate whether the shm|nm init hook has been called. What do you think ?

minsii replied:
I deleted this commit in the last push to keep this PR simple. I do not think it is an issue that we have to address in this PR.

hajimefu replied:
Okay.

minsii commented on src/mpid/ch4/shm/posix/posix_win.h:
```diff
@@ -409,27 +409,31 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_POSIX_mpi_win_allocate_hook(MPIR_Win * win)
     int mpi_errno = MPI_SUCCESS;
     MPIDI_POSIX_win_t *posix_win ATTRIBUTE((unused)) = NULL;
     MPIR_Comm *shm_comm_ptr = win->comm_ptr->node_comm;
+    bool mapfail_flag = false;
```
fixed

minsii commented on src/mpid/ch4/src/ch4r_symheap.h:
```diff
@@ -59,6 +59,12 @@
+    MPIDI_CH4I_SYMSHM_SUCCESS,
+    MPIDI_CH4I_SYMSHM_MAP_FAIL,
+    MPIDI_CH4I_SYMSHM_OTHER_FAIL
+};
```
added

minsii commented on src/mpid/ch4/src/ch4r_win.h:
```diff
@@ -1368,33 +1371,46 @@ static inline int MPIDI_CH4I_win_shm_alloc_impl(MPI_Aint size,
+                MPIDI_CH4U_WIN(win, mmap_addr) = NULL;
+            }
+
+            /* through error here if shm allocation is required but fails */
```
fixed

minsii commented on src/mpid/ch4/src/ch4r_win.h:
```diff
@@ -1277,7 +1278,8 @@ static inline int MPIDI_CH4R_mpi_win_attach(MPIR_Win * win, void *base, MPI_Aint
                                                 MPIR_Comm * comm_ptr,
-                                                void **base_ptr, MPIR_Win ** win_ptr)
+                                                void **base_ptr, MPIR_Win ** win_ptr,
+                                                bool shm_required)
```
I would think it is simply a yes or no option, thus is `true/false` sufficient ? The variable name already explains the meaning, and this is used only for the two window creation calls within this file.

hajimefu replied:
My point was that it was really hard to understand from the caller side.
```C
shm_alloc(size, ..., true);
```
or,
```C
shm_alloc(size, ..., SHM_REQUIRED);
```
I would think the latter is much easier to understand why it is doing, just by looking at the function invocation.

minsii replied:
OK, changed.

minsii commented on src/mpid/ch4/src/ch4_impl.h:
```diff
@@ -954,47 +954,73 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_CH4U_wait_am_acc(MPIR_Win * win, int target_r
-        MPIR_ERR_CHKANDJUMP(mpl_err, mpi_errno, MPI_ERR_OTHER, "**alloc_shar_mem");
+        if (mpl_err != MPL_SHM_SUCCESS) {
+            shm_fail_flag = true;
+            goto hnd_sync;
```
deleted the jump.

minsii commented on src/mpid/ch4/src/ch4r_symheap.h:
```diff
@@ -196,22 +202,29 @@ static inline void *MPIDI_CH4R_generate_random_addr(size_t size)
- * mapfail_flag is set to 1 if it is a mapping failure, otherwise it is 0. */
+ * map_result indicates the mapping result of the node. It can be success, map_fail
+ * or other fail. If it is map_fail, the caller can try it again with a different
+ * start address; if it is other_fail, it usually means no more shm segment can be
```
fixed

minsii commented:
test:jenkins/ch4/most

minsii commented:
The last push should fix most comments except the `init_hook|finalize_hook` issue.

minsii commented:
test:jenkins/ch4/most

minsii commented:
@hajimefu My last push should addressed all the comments. Could you please check again ?

Besides, the last push includes three new commits:
- `mpl: fix pthread_mutexattr_setpshared config check`
- `ch4/rma: fix symheap bug introduced when clean unused label`
    I found this issue after rebased with latests master.
- `ch4/rma: always reset mmap_addr|size when symheap fails`
   This will be merged into commit `ch4/rma: allow fallback when shm fails at win allocate` after your review.

I have locally tested possible mapping failure cases:
- `-np 2 -ppn 2 ./test/mpi/rma/window_allocation` 
   It first tries `MPIDI_CH4U_allocate_shm_segment`, if fails then call `malloc`
- `-np 2 -ppn 2 ./test/mpi/rma/window_noncontig_allocation` 
   It first tries `MPIDI_CH4R_get_shm_symheap`, if fails then try `MPIDI_CH4U_allocate_shm_segment`, if still fails then call `malloc`

Note that I can reproduce the "Too many open fils" (errno 24) mapping failure in the above tests only when `MPIR_CVAR_CH4_OFI_ENABLE_SHARED_CONTEXTS=1` is set.

minsii commented:
Test again after merging with latest master
test:jenkins/ch4/most

minsii commented:
Pushed again only for whitespace fix
